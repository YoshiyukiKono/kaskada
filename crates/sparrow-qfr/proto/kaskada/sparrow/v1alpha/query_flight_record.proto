syntax = "proto3";
package kaskada.sparrow.v1alpha;

import "google/protobuf/wrappers.proto";

option go_package = "gitlab.com/kaskada/kaskada/wren/gen/go/kaskada/sparrow/v1alpha";
option java_multiple_files = true;
option java_package = "com.kaskada.sparrow.v1alpha1";

// Metadata added at the beginning of the flight record file.
//
// Contains metadata on the query, plan, etc. useful for interepreting
// the flight record.
message FlightRecordHeader {
  // The version of the flight record format contained here.
  uint32 version = 1;

  // The request ID this flight record is for.
  string request_id = 2;

  // Build information about the Sparrow Binary.
  BuildInfo sparrow_build_info = 3;

  message BuildInfo {
    string sparrow_version = 1;
    string github_ref = 2;
    string github_sha = 3;
    string github_workflow = 4;
  }
}

message FlightRecord {
  oneof record {
    RegisterThread register_thread = 1;
    RegisterActivity register_activity = 2;
    RegisterMetric register_metric = 3;
    ReportActivity report_activity = 4;
    ReportMetrics report_metrics = 5;
  }

  message RegisterThread {
    uint32 thread_id = 1;
    string label = 2;
  }

  message RegisterActivity {
    uint32 activity_id = 1;

    string label = 2;

    // The ID of the parent activity.
    //
    // May be `None` / `null` to indicate a root activity.
    google.protobuf.UInt32Value parent_activity_id = 3;
  }

  message RegisterMetric {
    uint32 metric_id = 1;
    string label = 2;
    MetricKind kind = 3;

    // string description = 4;
    // string units = 5;

    enum MetricKind {
      METRIC_KIND_UNSPECIFIED = 0;

      METRIC_KIND_U64_GAUGE = 10;
      METRIC_KIND_I64_GAUGE = 11;
      METRIC_KIND_F64_GAUGE = 12;

      METRIC_KIND_U64_COUNTER = 20;
      METRIC_KIND_I64_COUNTER = 21;
      METRIC_KIND_F64_COUNTER = 22;
    }
  }

  message ReportActivity {
    uint32 activity_id = 1;
    uint32 thread_id = 2;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 3;

    // Wall clock time spent executing this event.
    //
    // This should be `0` for point-wise events.
    uint64 wall_duration_us = 4;

    // CPU time spent executing this event.
    uint64 cpu_duration_us = 5;

    // Metrics to report associated with the given activity.
    repeated MetricValue metrics = 6;
  }

  message ReportMetrics {
    uint32 thread_id = 1;

    // The time within the trace at which this event was recorded.
    uint64 wall_timestamp_us = 2;

    // Metrics to report against the given thread.
    repeated MetricValue metrics = 3;
  }
}

message MetricValue {
  uint32 metric_id = 1;

  oneof value {
    uint64 u64_value = 2;
    int64 i64_value = 3;
    double f64_value = 4;
  }
}