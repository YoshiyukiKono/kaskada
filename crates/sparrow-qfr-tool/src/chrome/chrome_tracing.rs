//! Contains structs and serializing corresponding to Chrome trace events.
//!
//! <https://aras-p.info/blog/2017/01/23/Chrome-Tracing-as-Profiler-Frontend/>
//!
//! <https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU>

#![allow(unused)]

use std::marker::PhantomData;
use std::num::{NonZeroU32, NonZeroUsize};

use hashbrown::HashMap;
use itertools::Itertools;
use serde::de::Visitor;
use serde::ser::{SerializeMap, SerializeStruct};
use serde::Serialize;
use smallvec::{smallvec, Array, SmallVec};
use sparrow_qfr::Times;

/// The outermost container for Chrome Tracing.
#[derive(Serialize, Eq, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub(super) struct Trace {
    trace_events: Vec<TraceEvent>,
    #[serde(default, skip_serializing_if = "DisplayTimeUnit::is_default")]
    display_time_unit: DisplayTimeUnit,
    #[serde(
        skip_serializing_if = "Vec::is_empty",
        serialize_with = "serialize_stack_frames"
    )]
    stack_frames: Vec<StackFrame>,
    #[serde(flatten)]
    other_data: HashMap<&'static str, &'static str>,
    #[serde(skip_serializing_if = "Args::is_empty")]
    metadata: Args<Arg>,
}

fn serialize_stack_frames<S>(frames: &[StackFrame], serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    let mut map = serializer.serialize_map(Some(frames.len()))?;
    for (index, frame) in frames.iter().enumerate() {
        map.serialize_entry(&(index + 1), frame)?;
    }
    map.end()
}

impl Trace {
    /// Create a new Trace with the given time unit.
    pub(super) fn new(display_time_unit: DisplayTimeUnit) -> Self {
        Self {
            trace_events: Vec::new(),
            display_time_unit,
            stack_frames: Vec::new(),
            other_data: HashMap::new(),
            metadata: Args::new(),
        }
    }

    pub fn metadata_mut(&mut self) -> &mut Args<Arg> {
        &mut self.metadata
    }

    /// Add a stack frame to the trace.
    pub fn add_stack_frame(&mut self, frame: StackFrame) -> StackFrameId {
        self.stack_frames.push(frame);
        let id = self.stack_frames.len() as u32;
        let id = NonZeroU32::new(id).expect("len should be non zero");
        StackFrameId(id)
    }

    pub fn get_stack_frame(&mut self, id: StackFrameId) -> &StackFrame {
        let index = id.0.get() - 1;
        &self.stack_frames[index as usize]
    }

    /// Add a process name for the process
    pub fn add_process_name(&mut self, process_id: u32, name: &'static str) {
        self.trace_events.push(TraceEvent {
            name: "process_name",
            categories: Categories::default(),
            event: Event::Metadata {
                args: Args(smallvec!(("name", Arg::StaticString(name)))),
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        })
    }

    pub fn add_thread_name(&mut self, process_id: u32, thread_id: u32, name: &'static str) {
        self.trace_events.push(TraceEvent {
            name: "thread_name",
            categories: Categories::default(),
            event: Event::Metadata {
                args: Args(smallvec!(("name", Arg::StaticString(name)))),
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id,
        })
    }

    pub fn add_event(
        &mut self,
        name: &'static str,
        wall_timestamp_us: u64,
        event: Event,
        process_id: u32,
        thread_id: u32,
    ) {
        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event,
            wall_timestamp_us,
            process_id,
            thread_id,
        })
    }

    pub fn add_object(
        &mut self,
        process_id: u32,
        name: &'static str,
        id: &'static str,
        args: Args<Arg>,
    ) {
        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event: Event::NewObject { id },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        });

        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event: Event::SnapshotObject {
                id,
                args: SnapshotArgs { snapshot: args },
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        });
    }
}

#[derive(Serialize, Eq, PartialEq, Debug)]
#[repr(transparent)]
pub(super) struct Category(&'static str);

impl From<&'static str> for Category {
    fn from(string: &'static str) -> Self {
        Category(string)
    }
}

#[derive(Eq, PartialEq, Debug, Default)]
#[repr(transparent)]
pub(super) struct Categories(SmallVec<[Category; 2]>);

impl Categories {
    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl<T> From<T> for Categories
where
    T: IntoIterator<Item = &'static str>,
{
    fn from(items: T) -> Self {
        Categories(items.into_iter().map(Category).collect())
    }
}

impl Serialize for Categories {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.collect_str(&self.0.iter().map(|cat| cat.0).format(","))
    }
}

#[derive(Eq, PartialEq, Debug)]
#[repr(transparent)]
pub(super) struct Args<T>(SmallVec<[(&'static str, T); 4]>);

impl<T> Args<T> {
    pub(super) fn new() -> Self {
        Args(SmallVec::new())
    }

    pub(super) fn with_capacity(n: usize) -> Self {
        Args(SmallVec::with_capacity(n))
    }

    pub(super) fn push(&mut self, name: &'static str, value: impl Into<T>) {
        self.0.push((name, value.into()))
    }

    pub(super) fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

#[derive(PartialEq, Eq, Debug)]
pub(super) enum Arg {
    Unsigned(u64),
    Signed(i64),
    StaticString(&'static str),
    String(String),
    ObjectRef(&'static str),
    Nested(Box<Args<Arg>>),
}

impl Serialize for Arg {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            Arg::Unsigned(n) => serializer.serialize_u64(*n),
            Arg::Signed(n) => serializer.serialize_i64(*n),
            Arg::StaticString(str) => serializer.serialize_str(str),
            Arg::String(str) => serializer.serialize_str(str),
            Arg::ObjectRef(object_id) => {
                let mut object_ref = serializer.serialize_struct("Reference", 1)?;
                object_ref.serialize_field("id_ref", &object_id)?;
                object_ref.end()
            }
            Arg::Nested(nested) => nested.serialize(serializer),
        }
    }
}

impl<T> From<SmallVec<[(&'static str, T); 4]>> for Args<T> {
    fn from(args: SmallVec<[(&'static str, T); 4]>) -> Self {
        Args(args)
    }
}

impl<T: Serialize> Serialize for Args<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.collect_map(self.0.iter().map(|(k, v)| (k, v)))
    }
}

#[derive(Serialize, Eq, PartialEq, Debug)]
struct TraceEvent {
    /// The name of the event, as displayed in Trace Viewer.
    name: &'static str,
    /// The event categories.
    ///
    /// This is a comma separated list of categories for the event.
    /// The categories can be used to hide events in the Trace Viewer UI.
    #[serde(rename = "cat", skip_serializing_if = "Categories::is_empty")]
    categories: Categories,
    /// The event type.
    ///
    /// This is a single character which changes depending
    /// on the type of event.
    #[serde(flatten)]
    event: Event,
    /// The tracing clock timestamp of the event.
    ///
    /// This should correspond to the elapsed time.
    ///
    /// The timestamps are provided at microsecond granularity.
    #[serde(rename = "ts")]
    wall_timestamp_us: u64,
    /// The process ID of the process that output this event.
    #[serde(rename = "pid")]
    process_id: u32,
    /// The thread ID of the process that output this event.
    #[serde(rename = "tid")]
    thread_id: u32,
}

/// Stack Frame ID.
#[repr(transparent)]
#[derive(Serialize, Eq, PartialEq, Debug, Hash, Clone, Copy)]
#[serde(transparent)]
pub(super) struct StackFrameId(NonZeroU32);

#[derive(Serialize, Eq, PartialEq, Debug)]
#[serde(tag = "ph")]
pub(super) enum Event {
    /// Indicates a complete event (start and end). Can reduce size of stack
    /// frame.
    #[serde(rename = "X")]
    Complete {
        /// Duration of the complete event, in microseconds.
        ///
        /// This should measure elapsed time.
        #[serde(rename = "dur")]
        wall_duration_us: u64,
        /// Thread clock duration of the complete event, in microseconds.
        ///
        /// This should measure CPU time.
        #[serde(rename = "tdur")]
        thread_duration_us: u64,
        #[serde(rename = "sf", skip_serializing_if = "Option::is_none")]
        stack_frame: Option<StackFrameId>,
        #[serde(rename = "esf", skip_serializing_if = "Option::is_none")]
        end_stack_frame: Option<StackFrameId>,
        #[serde(skip_serializing_if = "Args::is_empty")]
        args: Args<Arg>,
    },
    /// Indicates an instant event with no associated duration.
    #[serde(rename = "i")]
    Instant {
        /// The scope of the instant event.
        #[serde(rename = "s", skip_serializing_if = "EventScope::is_default")]
        scope: EventScope,
        #[serde(skip_serializing_if = "Args::is_empty")]
        args: Args<Arg>,
    },
    /// Assign metadata to a specific process or thread.
    ///
    /// These events require using special names for the event.
    ///
    /// If the name is `process_name`, then the argument `name` assigns the
    /// process name.
    ///
    /// If the name is `thread_name`, then the argument `name` assigns the
    /// thread name.
    #[serde(rename = "M")]
    Metadata { args: Args<Arg> },
    /// Adds a counter event.
    ///
    /// The names are scoped to the process. The thread ID is ignored.
    ///
    /// Each argument is a value for the metric.
    /// Multiple values are stacked at the point in time they occur.
    #[serde(rename = "C")]
    Counter { args: Args<i64> },

    /// Register the creation of an object with the given ID.
    #[serde(rename = "N")]
    NewObject { id: &'static str },
    /// Register the deletion of an object with the given ID.
    #[serde(rename = "D")]
    DestroyObject { id: &'static str },

    /// Register a snapshot of the given object.
    #[serde(rename = "O")]
    SnapshotObject {
        id: &'static str,
        args: SnapshotArgs,
    },

    /// Register the given object ID as the context for the current thread.
    #[serde(rename = "(")]
    ContextEnter { id: &'static str },

    /// Leave the given context for the current thread.
    #[serde(rename = ")")]
    ContextLeave { id: &'static str },
    // TODO: Flow events to associate information between threads?
}

#[derive(Serialize, PartialEq, Eq, Debug)]
pub(super) struct SnapshotArgs {
    snapshot: Args<Arg>,
}

/// The scope of an instant event.
///
/// Controls how tall the the event will be.
#[derive(Serialize, Eq, PartialEq, Debug)]
pub(super) enum EventScope {
    /// The event will be drawn from the top to bottom of the timeline.
    #[serde(rename = "g")]
    Global,
    /// The event will be drawn through all threads of a given process.
    #[serde(rename = "p")]
    Process,
    /// The event will be the height of a single thread.
    #[serde(rename = "t")]
    Thread,
}

impl Default for EventScope {
    fn default() -> Self {
        EventScope::Thread
    }
}

impl EventScope {
    fn is_default(&self) -> bool {
        self == &Self::default()
    }
}

#[derive(Serialize, Eq, PartialEq, Debug)]
pub(super) enum DisplayTimeUnit {
    #[serde(rename = "ms")]
    Millisecond,
    #[serde(rename = "ns")]
    Nanosecond,
}

impl Default for DisplayTimeUnit {
    fn default() -> Self {
        Self::Millisecond
    }
}

impl DisplayTimeUnit {
    fn is_default(&self) -> bool {
        self == &Self::default()
    }
}

#[derive(Serialize, Eq, PartialEq, Debug)]
pub(super) struct StackFrame {
    pub(super) category: &'static str,
    pub(super) name: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) parent: Option<StackFrameId>,
}

impl From<&'static str> for Arg {
    fn from(s: &'static str) -> Self {
        Self::StaticString(s)
    }
}

impl From<String> for Arg {
    fn from(s: String) -> Self {
        Self::String(s)
    }
}

impl From<u32> for Arg {
    fn from(n: u32) -> Self {
        Self::Unsigned(n as u64)
    }
}

impl From<i32> for Arg {
    fn from(n: i32) -> Self {
        Self::Signed(n as i64)
    }
}

impl From<u64> for Arg {
    fn from(n: u64) -> Self {
        Self::Unsigned(n)
    }
}

impl From<i64> for Arg {
    fn from(n: i64) -> Self {
        Self::Signed(n)
    }
}

impl From<Args<Arg>> for Arg {
    fn from(nested: Args<Arg>) -> Self {
        Self::Nested(Box::new(nested))
    }
}

#[cfg(test)]
mod tests {
    use std::convert::TryFrom;

    use smallvec::smallvec;

    use super::*;

    #[test]
    fn test_write_complete_event() {
        insta::assert_json_snapshot!(TraceEvent {
            name: "name",
            categories: Categories::default(),
            event: Event::Complete {
                wall_duration_us: 8736,
                thread_duration_us: 8730,
                stack_frame: None,
                end_stack_frame: None,
                args: Args::new(),
            },
            wall_timestamp_us: 57,
            process_id: 57,
            thread_id: 73,
        }, @r###"
        {
          "name": "name",
          "ph": "X",
          "dur": 8736,
          "tdur": 8730,
          "ts": 57,
          "pid": 57,
          "tid": 73
        }
        "###)
    }

    #[test]
    fn test_write_complete_event_with_categories() {
        insta::assert_json_snapshot!(TraceEvent {
            name: "name",
            categories: Categories::from(["apple", "banana"]),
            event: Event::Complete {
                wall_duration_us: 8736,
                thread_duration_us: 8730,
                stack_frame: None,
                end_stack_frame: None,
                args: Args::new(),
            },
            wall_timestamp_us: 57,
            process_id: 57,
            thread_id: 73,
        }, @r###"
        {
          "name": "name",
          "cat": "apple,banana",
          "ph": "X",
          "dur": 8736,
          "tdur": 8730,
          "ts": 57,
          "pid": 57,
          "tid": 73
        }
        "###)
    }

    // TODO: Add tests for the public API of trace.
    // TODO: Add a test with as stack frame.
    // TODO: Add test for counters, etc.
}
