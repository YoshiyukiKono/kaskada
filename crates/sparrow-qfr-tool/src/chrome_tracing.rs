//! Structs representing chrome traces.
//!
//! The serde-json format produces Chrome tracing JSON, as described in docs below.
//!
//! <https://aras-p.info/blog/2017/01/23/Chrome-Tracing-as-Profiler-Frontend/>
//!
//! <https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU>

mod args;
mod categories;
mod stack_frame;
mod trace;
mod trace_event;

pub(crate) use args::*;
pub(crate) use stack_frame::*;
pub(crate) use trace::*;
pub(crate) use trace_event::*;
