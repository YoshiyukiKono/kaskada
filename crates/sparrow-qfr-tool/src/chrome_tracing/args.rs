use serde::ser::SerializeStruct;
use serde::Serialize;
use smallvec::SmallVec;
use sparrow_qfr::kaskada::sparrow::v1alpha::metric_value;

#[derive(PartialEq, Debug)]
#[repr(transparent)]
pub(crate) struct Args(pub SmallVec<[(&'static str, Arg); 4]>);

impl Args {
    pub fn new() -> Self {
        Args(SmallVec::new())
    }

    pub fn push(&mut self, name: &'static str, value: impl Into<Arg>) {
        self.0.push((name, value.into()))
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

#[derive(PartialEq, Debug)]
pub(crate) enum Arg {
    Unsigned(u64),
    Signed(i64),
    Float(f64),
    StaticString(&'static str),
    String(String),
    #[allow(dead_code)]
    ObjectRef(&'static str),
    Nested(Box<Args>),
}

impl Serialize for Arg {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            Arg::Unsigned(n) => serializer.serialize_u64(*n),
            Arg::Signed(n) => serializer.serialize_i64(*n),
            Arg::Float(n) => serializer.serialize_f64(*n),
            Arg::StaticString(str) => serializer.serialize_str(str),
            Arg::String(str) => serializer.serialize_str(str),
            Arg::ObjectRef(object_id) => {
                let mut object_ref = serializer.serialize_struct("Reference", 1)?;
                object_ref.serialize_field("id_ref", &object_id)?;
                object_ref.end()
            }
            Arg::Nested(nested) => nested.serialize(serializer),
        }
    }
}

impl From<SmallVec<[(&'static str, Arg); 4]>> for Args {
    fn from(args: SmallVec<[(&'static str, Arg); 4]>) -> Self {
        Args(args)
    }
}

impl Serialize for Args {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.collect_map(self.0.iter().map(|(k, v)| (k, v)))
    }
}

impl From<&'static str> for Arg {
    fn from(s: &'static str) -> Self {
        Self::StaticString(s)
    }
}

impl From<String> for Arg {
    fn from(s: String) -> Self {
        Self::String(s)
    }
}

impl From<u32> for Arg {
    fn from(n: u32) -> Self {
        Self::Unsigned(n as u64)
    }
}

impl From<i32> for Arg {
    fn from(n: i32) -> Self {
        Self::Signed(n as i64)
    }
}

impl From<u64> for Arg {
    fn from(n: u64) -> Self {
        Self::Unsigned(n)
    }
}

impl From<i64> for Arg {
    fn from(n: i64) -> Self {
        Self::Signed(n)
    }
}

impl From<Args> for Arg {
    fn from(nested: Args) -> Self {
        Self::Nested(Box::new(nested))
    }
}

impl From<metric_value::Value> for Arg {
    fn from(value: metric_value::Value) -> Self {
        match value {
            metric_value::Value::U64Value(n) => Arg::Unsigned(n),
            metric_value::Value::I64Value(n) => Arg::Signed(n),
            metric_value::Value::F64Value(n) => Arg::Float(n),
        }
    }
}
