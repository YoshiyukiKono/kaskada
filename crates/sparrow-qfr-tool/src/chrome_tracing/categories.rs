use itertools::Itertools;
use serde::Serialize;
use smallvec::SmallVec;

/// Wrapper around a static string used as a category.
#[derive(Serialize, Eq, PartialEq, Debug)]
#[repr(transparent)]
pub(crate) struct Category(&'static str);

impl From<&'static str> for Category {
    fn from(string: &'static str) -> Self {
        Category(string)
    }
}

/// Vector of categories.
#[derive(Eq, PartialEq, Debug, Default)]
#[repr(transparent)]
pub(crate) struct Categories(SmallVec<[Category; 2]>);

impl Categories {
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl<T> From<T> for Categories
where
    T: IntoIterator<Item = &'static str>,
{
    fn from(items: T) -> Self {
        Categories(items.into_iter().map(Category).collect())
    }
}

impl Serialize for Categories {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.collect_str(&self.0.iter().map(|cat| cat.0).format(","))
    }
}
