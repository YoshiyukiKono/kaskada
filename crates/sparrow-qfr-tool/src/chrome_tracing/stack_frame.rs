use serde::Serialize;

#[derive(Serialize, Eq, PartialEq, Debug)]
pub(crate) struct StackFrame {
    #[serde(skip_serializing)]
    pub(crate) id: StackFrameId,
    #[serde(skip_serializing_if = "str::is_empty")]
    pub(crate) category: &'static str,
    pub(crate) name: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) parent: Option<StackFrameId>,
}

/// Stack Frame ID.
#[repr(transparent)]
#[derive(Serialize, Eq, PartialEq, Debug, Hash, Clone, Copy)]
#[serde(transparent)]
pub(crate) struct StackFrameId(pub u32);
