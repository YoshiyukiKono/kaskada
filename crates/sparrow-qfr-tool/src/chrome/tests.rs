use sparrow_qfr::kaskada::sparrow::v1alpha::flight_record::{
    Record, RegisterThread, ReportActivity, ReportMetrics,
};
use sparrow_qfr::kaskada::sparrow::v1alpha::flight_record_header::register_metric::MetricKind;
use sparrow_qfr::kaskada::sparrow::v1alpha::flight_record_header::{
    BuildInfo, RegisterActivity, RegisterMetric,
};
use sparrow_qfr::kaskada::sparrow::v1alpha::{
    metric_value, FlightRecord, FlightRecordHeader, MetricValue,
};

use crate::chrome::conversion::Conversion;
use crate::chrome_tracing::Trace;

fn run_test_ok(
    activities: Vec<RegisterActivity>,
    metrics: Vec<RegisterMetric>,
    records: Vec<Record>,
) -> Trace {
    let header = FlightRecordHeader {
        version: sparrow_qfr::QFR_VERSION,
        request_id: "request_id".to_owned(),
        sparrow_build_info: Some(BuildInfo {
            sparrow_version: "test_version".to_owned(),
            github_ref: "".to_owned(),
            github_sha: "".to_owned(),
            github_workflow: "".to_owned(),
        }),
        activities,
        metrics,
    };

    let conversion = Conversion::new(&header);
    let records =
        fallible_iterator::convert::<_, error_stack::Report<sparrow_qfr::io::reader::Error>, _>(
            records
                .into_iter()
                .map(|record| FlightRecord {
                    record: Some(record),
                })
                .map(Ok),
        );
    conversion
        .convert_all(records)
        .expect("successful conversion")
}

#[test]
fn test_conversion_empty_trace() {
    let activities = vec![];
    let metrics = vec![];
    let records = vec![];
    insta::assert_json_snapshot!(run_test_ok(activities, metrics, records));
}

#[test]
fn test_conversion_some_activities() {
    let activities = vec![
        RegisterActivity {
            activity_id: 17,
            label: "child1".to_owned(),
            parent_activity_id: Some(18),
        },
        RegisterActivity {
            activity_id: 18,
            label: "root".to_owned(),
            parent_activity_id: None,
        },
        RegisterActivity {
            activity_id: 19,
            label: "child2".to_owned(),
            parent_activity_id: Some(18),
        },
    ];
    let metrics = vec![];

    let thread_id = 11;
    let records = vec![
        Record::RegisterThread(RegisterThread {
            thread_id,
            label: "operation 0".to_owned(),
        }),
        Record::ReportActivity(ReportActivity {
            activity_id: 19,
            thread_id,
            wall_timestamp_us: 1_000_000,
            wall_duration_us: 2_573,
            cpu_duration_us: 1_172,
            metrics: vec![],
        }),
    ];
    insta::assert_json_snapshot!(run_test_ok(activities, metrics, records));
}

#[test]
fn test_conversion_activity_metrics() {
    let activities = vec![RegisterActivity {
        activity_id: 18,
        label: "root".to_owned(),
        parent_activity_id: None,
    }];
    let metrics = vec![RegisterMetric {
        metric_id: 87,
        label: "gauge_metric".to_owned(),
        kind: MetricKind::U64Gauge as i32,
    }];

    let thread_id = 11;
    let records = vec![
        Record::RegisterThread(RegisterThread {
            thread_id,
            label: "operation 0".to_owned(),
        }),
        Record::ReportActivity(ReportActivity {
            activity_id: 18,
            thread_id,
            wall_timestamp_us: 1_000_000,
            wall_duration_us: 2_573,
            cpu_duration_us: 1_172,
            metrics: vec![MetricValue {
                metric_id: 87,
                value: Some(metric_value::Value::U64Value(18)),
            }],
        }),
    ];
    insta::assert_json_snapshot!(run_test_ok(activities, metrics, records));
}

#[test]
fn test_conversion_root_metrics() {
    let activities = vec![RegisterActivity {
        activity_id: 18,
        label: "root".to_owned(),
        parent_activity_id: None,
    }];
    let metrics = vec![RegisterMetric {
        metric_id: 87,
        label: "gauge_metric".to_owned(),
        kind: MetricKind::U64Gauge as i32,
    }];

    let thread_id = 11;
    let records = vec![
        Record::RegisterThread(RegisterThread {
            thread_id,
            label: "operation 0".to_owned(),
        }),
        Record::ReportMetrics(ReportMetrics {
            thread_id,
            wall_timestamp_us: 1_000_000,
            metrics: vec![MetricValue {
                metric_id: 87,
                value: Some(metric_value::Value::U64Value(18)),
            }],
        }),
    ];
    insta::assert_json_snapshot!(run_test_ok(activities, metrics, records));
}
