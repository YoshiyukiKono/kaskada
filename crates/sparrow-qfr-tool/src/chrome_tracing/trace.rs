use std::fs::File;
use std::io::{BufWriter, Write};

use crate::chrome_tracing::args::{Arg, Args};
use crate::chrome_tracing::categories::Categories;
use crate::chrome_tracing::stack_frame::StackFrame;
use crate::chrome_tracing::trace_event::{Event, SnapshotArgs};
use crate::error::Error;

use super::trace_event::TraceEvent;
use error_stack::{IntoReport, ResultExt};
use hashbrown::HashMap;
use serde::ser::SerializeMap;
use serde::Serialize;
use smallvec::smallvec;

/// The outermost container for Chrome Tracing.
#[derive(Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Trace {
    trace_events: Vec<TraceEvent>,
    #[serde(default, skip_serializing_if = "DisplayTimeUnit::is_default")]
    display_time_unit: DisplayTimeUnit,
    #[serde(
        skip_serializing_if = "Vec::is_empty",
        serialize_with = "serialize_stack_frames"
    )]
    stack_frames: Vec<StackFrame>,
    #[serde(flatten)]
    other_data: HashMap<&'static str, &'static str>,
    #[serde(skip_serializing_if = "Args::is_empty")]
    metadata: Args,
}

fn serialize_stack_frames<S>(frames: &[StackFrame], serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    let mut map = serializer.serialize_map(Some(frames.len()))?;
    for frame in frames.iter() {
        map.serialize_entry(&frame.id, frame)?;
    }
    map.end()
}

impl Trace {
    /// Create a new Trace with the given time unit.
    pub fn new(display_time_unit: DisplayTimeUnit) -> Self {
        Self {
            trace_events: Vec::new(),
            display_time_unit,
            stack_frames: Vec::new(),
            other_data: HashMap::new(),
            metadata: Args::new(),
        }
    }

    pub fn metadata_mut(&mut self) -> &mut Args {
        &mut self.metadata
    }

    /// Add a stack frame to the trace.
    pub fn add_stack_frame(&mut self, frame: StackFrame) {
        self.stack_frames.push(frame);
    }

    /// Add a process name for the process
    pub fn add_process_name(&mut self, process_id: u32, name: &'static str) {
        self.trace_events.push(TraceEvent {
            name: "process_name",
            categories: Categories::default(),
            event: Event::Metadata {
                args: Args(smallvec!(("name", Arg::StaticString(name)))),
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        })
    }

    #[allow(dead_code)]
    pub fn add_thread_name(&mut self, process_id: u32, thread_id: u32, name: &'static str) {
        self.trace_events.push(TraceEvent {
            name: "thread_name",
            categories: Categories::default(),
            event: Event::Metadata {
                args: Args(smallvec!(("name", Arg::StaticString(name)))),
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id,
        })
    }

    pub fn add_event(
        &mut self,
        name: &'static str,
        wall_timestamp_us: u64,
        event: Event,
        process_id: u32,
        thread_id: u32,
    ) {
        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event,
            wall_timestamp_us,
            process_id,
            thread_id,
        })
    }

    #[allow(dead_code)]
    pub fn add_object(
        &mut self,
        process_id: u32,
        name: &'static str,
        id: &'static str,
        args: Args,
    ) {
        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event: Event::NewObject { id },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        });

        self.trace_events.push(TraceEvent {
            name,
            categories: Categories::default(),
            event: Event::SnapshotObject {
                id,
                args: SnapshotArgs { snapshot: args },
            },
            wall_timestamp_us: 0,
            process_id,
            thread_id: 0,
        });
    }

    pub fn write_to(&self, path: &std::path::Path) -> error_stack::Result<(), Error> {
        let output = File::create(path)
            .into_report()
            .change_context(Error::Internal)?;
        let mut buffer = BufWriter::new(output);
        serde_json::to_writer(&mut buffer, self)
            .into_report()
            .change_context(Error::Internal)?;
        buffer
            .flush()
            .into_report()
            .change_context(Error::Internal)?;

        Ok(())
    }
}

/// Time unit to display on traces.
#[derive(Serialize, Eq, PartialEq, Debug)]
pub(crate) enum DisplayTimeUnit {
    #[serde(rename = "ms")]
    Millisecond,
    #[serde(rename = "ns")]
    #[allow(dead_code)]
    Nanosecond,
}

impl Default for DisplayTimeUnit {
    fn default() -> Self {
        Self::Millisecond
    }
}

impl DisplayTimeUnit {
    fn is_default(&self) -> bool {
        self == &Self::default()
    }
}
