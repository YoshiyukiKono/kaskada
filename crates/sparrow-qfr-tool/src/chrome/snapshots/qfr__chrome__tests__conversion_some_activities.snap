---
source: crates/sparrow-qfr-tool/src/chrome/tests.rs
expression: "run_test_ok(activities, metrics, records)"
---
{
  "traceEvents": [
    {
      "name": "process_name",
      "ph": "M",
      "args": {
        "name": "operation 0"
      },
      "ts": 0,
      "pid": 11,
      "tid": 0
    },
    {
      "name": "child2",
      "ph": "X",
      "dur": 2573,
      "tdur": 1172,
      "sf": 19,
      "ts": 1000000,
      "pid": 11,
      "tid": 0
    }
  ],
  "stackFrames": {
    "17": {
      "name": "child1",
      "parent": 18
    },
    "18": {
      "name": "root"
    },
    "19": {
      "name": "child2",
      "parent": 18
    }
  },
  "metadata": {
    "request_id": "request_id",
    "build_info": {
      "sparrow_version": "test_version",
      "github_ref": "",
      "github_sha": "",
      "github_workflow": ""
    }
  }
}
